src/main/java/ca/nexapp/core/dates/DateTimeRange.java

```diff
@@ -0,0 +1,104 @@
+package ca.nexapp.core.dates;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.Objects;
+import java.util.Optional;
+
+public class DateTimeRange {
+
+    private final ZonedDateTime from;
+    private final ZonedDateTime to;
+
+    private DateTimeRange(ZonedDateTime from, ZonedDateTime to) throws InvalidDateRangeException {
+        if (from.isAfter(to)) {
+            throw new InvalidDateRangeException(from, to);
+        }
+
+        this.from = from;
+        this.to = to;
+    }
+
+    public boolean isOverlapping(DateTimeRange other) {
+        return !(from.isAfter(other.to) || to.isBefore(other.from));
+    }
+
+    public Optional<DateTimeRange> findOverlappingPeriod(DateTimeRange other) {
+        if (!isOverlapping(other)) {
+            return Optional.empty();
+        }
+
+        ZonedDateTime startDate = from.compareTo(other.from) > 0 ? from : other.from;
+        ZonedDateTime endDate = to.compareTo(other.to) < 0 ? to : other.to;
+        DateTimeRange dateRange = DateTimeRange.of(startDate, endDate);
+        return Optional.of(dateRange);
+    }
+
+    public Duration asDuration() {
+        return Duration.between(from, to);
+    }
+
+    public ZoneId getZone() {
+        return from.getZone();
+    }
+
+    public ZonedDateTime getFrom() {
+        return from;
+    }
+
+    public ZonedDateTime getTo() {
+        return to;
+    }
+
+    public boolean includes(Optional<ZonedDateTime> date) {
```

Comment:
which situations do you see this method be used? wouldn't it be better to let the callee do that by itself?